pkg/virt-handler/device-manager/device_plugin_base_test.go

```diff
@@ -0,0 +1,82 @@
+package device_manager
+
+import (
+	"errors"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"google.golang.org/grpc"
+
+	pluginapi "kubevirt.io/kubevirt/pkg/virt-handler/device-manager/deviceplugin/v1beta1"
+)
+
+type fakeListAndWatchServer struct {
+	listAndWatchResponseError error
+	count                     int
+	grpc.ServerStream
+}
+
+func (f *fakeListAndWatchServer) setlistAndWatchResponseError(err error) {
+	f.listAndWatchResponseError = err
+}
+
+func (f *fakeListAndWatchServer) Send(m *pluginapi.ListAndWatchResponse) error {
+	f.count++
+	return f.listAndWatchResponseError
+}
+
+var _ = Describe("Device plugin base", func() {
+	Context("List and Watch", func() {
+		var (
+			fakeServer   *fakeListAndWatchServer
+			devicePlugin *DevicePluginBase
+			stop         chan struct{}
+		)
+		BeforeEach(func() {
+			stop = make(chan struct{})
+			fakeServer = &fakeListAndWatchServer{
+				count: 0,
+			}
+			devicePlugin = &DevicePluginBase{
+				stop: stop,
+			}
+		})
+		AfterEach(func() {
+			close(stop)
+		})
+
+		It("should return an error when the send the request to the server fails", func() {
+			errExpected := errors.New("error")
+			fakeServer.setlistAndWatchResponseError(errExpected)
+			err := devicePlugin.ListAndWatch(nil, fakeServer)
+			Expect(err).To(Equal(errExpected))
+		})
+
+		It("should gracefully stop the device plugin", func() {
+			errChan := make(chan error, 1)
+			go func(errChan chan error) {
+				errChan <- devicePlugin.ListAndWatch(nil, fakeServer)
+			}(errChan)
+			stop <- struct{}{}
+			Eventually(func() error {
+				return (<-errChan)
+			}, 5*time.Second).ShouldNot(HaveOccurred())
+		})
+
+		It("should send the device infomation to the server", func() {
+			errChan := make(chan error, 1)
+			devicePlugin.devs = []*pluginapi.Device{
+				&pluginapi.Device{}, &pluginapi.Device{}, &pluginapi.Device{},
+			}
+			go func(errChan chan error) {
+				errChan <- devicePlugin.ListAndWatch(nil, fakeServer)
+			}(errChan)
+			stop <- struct{}{}
+			Eventually(func() error {
+				return (<-errChan)
+			}, 5*time.Second).ShouldNot(HaveOccurred())
+			Expect(fakeServer.count).To(Equal(3))
```

Comment:
3 seems like magic number, can we add small explanation?

Reply:
We have added 3 Device plugin a couple of line above, but I can add a comment

Reply:
First of, the 3 does not come from 3 device plugins, as looking at the code, it counds the number of times `Send` is called. This is 2-3 times in this case.
It took me a bit to figure out, but this is also the reason why this test can be flaky/failing on s390x. Since one call to `Send` happens in the endless loop, it is in a race with the `stop` channel to see which is triggered first.
At least that is my interpretation of what might be happening with the s390x unit-test. Our timings seem to always be a bit off since we are using the `-race` flag.